File: algorithm_study/xinghuaOj/10001种树.py
"""
9
4 
1 4 2 
4 6 1 
8 9 2 
3 5 2 


[1, 4, 2, 0]
[3, 5, 2, 0]
[4, 6, 1, 0]
[8, 9, 2, 0]
"""
treeNums = int(input())
hopeNums = int(input())
targetList = []

for i in range(hopeNums):
    targetList.append(list(map(int, input().split())))
targetList.sort(key=lambda x: (x[1]))  #从小到大排
roadList = [False for n in range(0, targetList[hopeNums - 1][1] + 1)]  #路段似乎得建大点 因为遍历的时候没0点
treeCount = 0
for it in targetList:
    # print(it)
    if (it[2] > 0):  #需要目标还没完成 就继续种树
        for j in range(it[1], it[0] - 1, -1):
            if (roadList[j]):
                it[2] -= 1  #这里目标值不减的话会误报

        for j in range(it[1], it[0] - 1, -1):
            # print("j:", j)
            if (roadList[j] == False):
                """
                倒着遍历每个区间 如果已经种的树比需要种的少 
                且 如果标记为没种树 
                """
                if (it[2] == 0):
                    break
                roadList[j] = True  #标记已种树
                # print("在 roadlist", j, "位置种树")
                it[2] -= 1  #目标值-=1
                treeCount += 1

# for it in targetList:
#     print(it)
print(treeCount)

File: algorithm_study/xinghuaOj/10001种树.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;
const int N = 1e6 + 10;
class target
{
  public:
    int begin;
    int end;
    int hope;
};
void f(target t)
{
    cout << "[" << t.begin << "," << t.end << "," << t.hope << "]" << endl;
}
bool cmp(target a, target b) { return a.end < b.end; }
int main()
{
    // input
    int treeCount = 0, hopeNums = 0, treeNums = 0;
    target targetList[50001];
    bool roadList[N] = {false};
    cin >> treeNums;
    cin >> hopeNums;
    for (int i = 0; i < hopeNums; i++)
    {
        cin >> targetList[i].begin >> targetList[i].end >> targetList[i].hope;
    }
    // sort
    sort(targetList, targetList + hopeNums - 1, cmp);
    // for_each(targetList.begin(), targetList.end(), f);
    for (auto it : targetList)
    {
        //cout << "[" << it.begin << "," << it.end << "," << it.hope << "]" << endl;
        if (it.hope > 0)
        {
            for (auto j = it.begin; j <= it.end; j++)
            {
                //先扫一次 看看该节点内有没被标记的
                //   cout << j << ",";
                if (roadList[j])
                    it.hope--;
            }
            // cout << endl;
            for (auto j = it.end; j >= it.begin; j--)
            {
                if (!roadList[j])
                {
                    if (it.hope < 1)
                        break;
                    roadList[j] = true;
                    //cout << "在 roadlist" << j << "位置种树" << endl;
                    it.hope--;
                    treeCount++;
                }
            }
        }
    }
    cout << treeCount;
    return 0;
}